**Importing the Required Libraries**

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

**Importing the Required Dataset**

In [2]:
dataset = pd.read_csv('Data.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

**Splitting the dataset**

In [3]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

**Multiple Linear Regression**

In [4]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

LinearRegression()

In [5]:
y_pred = regressor.predict(X_test)
np.set_printoptions(precision=2)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[431.43 431.23]
 [458.58 460.01]
 [462.77 461.14]
 ...
 [473.75 468.9 ]
 [475.44 473.9 ]
 [442.02 437.07]]


In [6]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.9323789104734466

**Polynomial Linear Regression**

In [8]:
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X_train)
lin_reg = LinearRegression()
lin_reg.fit(X_poly, y_train)


LinearRegression()

In [10]:
y_pred1 = lin_reg.predict(poly_reg.transform(X_test))
np.set_printoptions(precision=2)
print(np.concatenate((y_pred.reshape(len(y_pred1),1), y_test.reshape(len(y_test),1)),1))

[[433.98 431.23]
 [457.97 460.01]
 [460.69 461.14]
 ...
 [470.31 468.9 ]
 [477.79 473.9 ]
 [440.03 437.07]]


In [11]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred1)

0.9459731906012439

**Random Forest**

In [13]:
from sklearn.ensemble import RandomForestRegressor
random_regressor = RandomForestRegressor(n_estimators = 10, random_state = 0)
random_regressor.fit(X_train, y_train)

RandomForestRegressor(n_estimators=10, random_state=0)

In [14]:
y_pred2 = random_regressor.predict(X_test)
np.set_printoptions(precision=2)
print(np.concatenate((y_pred2.reshape(len(y_pred2),1), y_test.reshape(len(y_test),1)),1))

[[434.54 431.23]
 [458.79 460.01]
 [462.62 461.14]
 ...
 [470.44 468.9 ]
 [475.13 473.9 ]
 [439.56 437.07]]


In [15]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred2)

0.960853142550123

**Decision Tree Regression**

In [16]:
from sklearn.tree import DecisionTreeRegressor
decision_regressor = DecisionTreeRegressor(random_state = 0)
decision_regressor.fit(X_train, y_train)

DecisionTreeRegressor(random_state=0)

In [17]:
y_pred3 = decision_regressor.predict(X_test)
np.set_printoptions(precision=2)
print(np.concatenate((y_pred3.reshape(len(y_pred3),1), y_test.reshape(len(y_test),1)),1))

[[432.98 431.23]
 [454.67 460.01]
 [463.6  461.14]
 ...
 [475.89 468.9 ]
 [474.15 473.9 ]
 [439.99 437.07]]


In [18]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred3)

0.9229733351311707

**Support Vector Regression**

In [22]:
y = y.reshape(len(y),1)

In [23]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [25]:
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)

In [26]:
from sklearn.svm import SVR
regressor = SVR(kernel = 'rbf')
regressor.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVR()

In [28]:
y_pred4 = sc_y.inverse_transform(regressor.predict(sc_X.transform(X_test)).reshape(-1, 1)) # Reshape the predictions to a 2D array
np.set_printoptions(precision=2)
print(np.concatenate((y_pred4.reshape(len(y_pred4),1), y_test.reshape(len(y_test),1)),1))

[[457.54 431.23]
 [457.54 460.01]
 [457.54 461.14]
 ...
 [457.54 473.26]
 [457.54 438.  ]
 [457.54 463.28]]


In [29]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred4)

-0.02931874555836056

***Based on our regression analysis, the Random Forest model achieved the highest performance with an R² score of 0.96, making it the most accurate model for predicting the Energy rate by using Ambient Temperature and other factors***